File: Algorithm/Python/486. Predict the Winner.py
# -*- coding: utf-8 -*-
# @Author: WuLC
# @Date:   2017-02-10 11:11:15
# @Last modified by:   WuLC
# @Last Modified time: 2017-02-10 11:11:34
# @Email: liangchaowu5@gmail.com


# recursive method , AC
class Solution(object):
    def PredictTheWinner(self, nums):
        """
        :type nums: List[int]
        :rtype: bool
        """
        return self.firstWin(0, 0, nums, 0, len(nums)-1)
    
    def firstWin(self, s1, s2, nums, start, end):
        """return true if the first person to pick the number finally win"""
        if start > end:
            return True if s1>=s2 else False
        return not (self.secondWin(s1+nums[start], s2, nums, start+1, end) and self.secondWin(s1+nums[end], s2, nums, start, end-1))
        
    def secondWin(self, s1, s2, nums, start, end):
        """return true if the second person to pick the number finally win"""
        if start > end:
            return True if s1 < s2 else False
        return not (self.firstWin(s1, s2+nums[start], nums, start+1, end) and self.firstWin(s1, s2+nums[end], nums, start, end-1))

File: Algorithm/Java/486. Predict the Winner.java
/**
* Author: WuLC
* Date:   2017-05-24 12:50:26
* Last modified by:   WuLC
* Last Modified time: 2017-05-24 12:52:40
* Email: liangchaowu5@gmail.com
*/


// recursive 
// P1Win represent whether P1 can win with the game with the left numbers
// P2Win represent whether P2 can win with the game with the left numbers
public class Solution 
{
    public boolean PredictTheWinner(int[] nums) 
    {
        return P1Win(nums, 0, nums.length - 1, 0, 0);
    }
    
    public boolean P1Win(int[] nums, int start, int end, int sum1, int sum2)
    {
        if (start > end)
        {
            if(sum1 >= sum2) return true;
            else return false;
        }
        if (P2Win(nums, start+1, end, sum1+nums[start], sum2) && P2Win(nums, start, end-1, sum1+nums[end], sum2)) return false;
        else return true;
    }
    
    public boolean P2Win(int[] nums, int start, int end, int sum1, int sum2)
    {
        if (start > end)
        {
            if(sum2 > sum1) return true;
            else return false;
        }
        if (P1Win(nums, start+1, end, sum1, sum2+nums[start]) && P1Win(nums, start, end-1, sum1, sum2+nums[end])) return false;
        else return true;
    }
}